reference/src/main/scala/akka/stream/alpakka/reference/model.scala

```diff
@@ -0,0 +1,139 @@
+ */
+
+package akka.stream.alpakka.reference
+
+// rename Java imports if the name clashes with the Scala name
+import java.lang.{Long => JavaLong}
+import java.util.{List => JavaList, Map => JavaMap}
+
+import akka.util.ByteString
+
+import scala.annotation.varargs
+import scala.collection.immutable
+import akka.japi.Pair
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Use "Read" in message data types to signify that the message was read from outside.
+ */
+final class ReferenceReadMessage private (
+    val data: immutable.Seq[ByteString] = immutable.Seq.empty,
+    val bytesRead: Try[Int] = Success(0)
+) {
+  def withData(data: immutable.Seq[ByteString]): ReferenceReadMessage =
+    copy(data = data)
+
+  def withBytesRead(bytesRead: Try[Int]): ReferenceReadMessage =
+    copy(bytesRead = bytesRead)
+
+  /**
+   * Java API
+   *
+   * If the model class is meant to be also consumed from the user API,
+   * but the attribute class is Scala specific, create getter for Java API.
+   */
+  def getData(): JavaList[ByteString] = {
+    import scala.collection.JavaConverters._
+    data.asJava
+  }
+
+  /**
+   * Java API
+   *
+   * If the model class is scala.util.Try, then two getters should be created.
+   * One for getting the value, and another for getting the exception.
+   *
+   * Should return null if the Try contains a Failure.
+   */
+  def getBytesRead(): Integer = bytesRead match {
+    case Success(bytesRead) => bytesRead
+    case Failure(_) => null
+  }
+
+  /**
+   * Java API
+   *
+   * Return the exception if the Try contains a Failure, otherwise return a null.
+   */
+  def getBytesReadFailure(): Throwable = bytesRead match {
+    case Success(_) => null
+    case Failure(ex) => ex
+  }
+
+  private def copy(data: immutable.Seq[ByteString] = data, bytesRead: Try[Int] = bytesRead) =
+    new ReferenceReadMessage(data, bytesRead)
+}
+
+object ReferenceReadMessage {
+  def apply(): ReferenceReadMessage = new ReferenceReadMessage()
+}
+
+/**
+ * Use "Write" in message data types to signify that the messages is to be written to outside.
+ */
+final class ReferenceWriteMessage private (
+    val data: immutable.Seq[ByteString] = immutable.Seq.empty,
+    val metrics: Map[String, Long] = Map.empty
+) {
+  def withData(data: immutable.Seq[ByteString]): ReferenceWriteMessage =
+    copy(data = data)
+
+  def withMetrics(metrics: Map[String, Long]): ReferenceWriteMessage =
+    copy(metrics = metrics)
+
+  /**
+   * Java API
+   *
+   * When settings class has an attribute of Scala collection type,
+   * use varargs annotation to generate a Java API varargs method.
+   */
+  @varargs def withData(data: ByteString*): ReferenceWriteMessage =
+    copy(data = data.to[immutable.Seq])
+
+  /**
+   * Java API
+   *
+   * Java setter needs to take Java Long class and convert to Scala Long.
+   */
+  @varargs def withMetrics(metrics: Pair[String, JavaLong]*): ReferenceWriteMessage =
```

Comment:
Even add `def withMetrics(key: String, value: JavaLong)`?

Reply:
Do you think it is worth it? The equivalent of this would be `withMetrics(Pair.create(key, value))`

Reply:
It would let Java users get away without learning about Akka's `Pair`.

Reply:
Okay. Then lets change this setter and instead of `Pair` let it take `Map.Entry`.

Reply:
The ugly thing is the warning "Unchecked generics array creation" which needs to be oppressed by `@SuppressWarnings("unchecked")`.